examples/notebooks/beam-ml/rag_usecase/opensearch_connector.py

```diff
@@ -0,0 +1,383 @@
+Example usage::
+
+  pipeline | WriteToOpensearch(host='localhost',
+                          port=6379,
+                          batch_size=100)
+
+
+No backward compatibility guarantees. Everything in this module is experimental.
+"""
+
+class InsertEmbeddingInOpenSearch(PTransform):
+    """WriteToOpensearch is a ``PTransform`` that writes a ``PCollection`` of
+    key, value tuple or 2-element array into a Opensearch server.
+    """
+
+    def __init__(self, 
+                 host:str, 
+                 port:int, 
+                 username:str, 
+                 password:str,
+                 batch_size:int=100, 
+                 embedded_columns:list=[]
+        ):
+        """
+
+        Args:
+        host (str, ValueProvider): The Opensearch host
+        port (int, ValueProvider): The Opensearch port
+        batch_size(int, ValueProvider): Number of key, values pairs to write at once
+        embedded_columns (list, ValueProvider): list of column whose embedding needs to be generated
+
+        Returns:
+        :class:`~apache_beam.transforms.ptransform.PTransform`
+
+        """
+
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+        self.batch_size = batch_size
+        self.embedded_columns = embedded_columns
+
+    def expand(self, pcoll):
+        return pcoll \
+               | "Reshuffle for Embedding in Opensearch Insert" >> Reshuffle() \
+               | "Write `Embeddings` to Opensearch" >> beam.ParDo(_WriteEmbeddingInOpenSearchFn(self.host,
+                                          self.port,
+                                          self.username,
+                                          self.password,
+                                          self.batch_size,
+                                          self.embedded_columns))
+
+
+class _WriteEmbeddingInOpenSearchFn(DoFn):
+    """Abstract class that takes in Opensearch  credentials 
+    to connect to Opensearch DB
+    """
+    
+    def __init__(self, 
+                 host:str, 
+                 port:int, 
+                 username:str,
+                 password:str,
+                 batch_size:int=100,
+                 embedded_columns:list=[]):
+        self.host = host
+        self.port = port
+        self.username = username
+        self.password = password
+        self.batch_size = batch_size
+        self.embedded_columns = embedded_columns
+
+        self.batch_counter = 0
+        self.batch = list()
+
+    def finish_bundle(self):
+        self._flush()
+
+    def process(self, element, *args, **kwargs):
+        self.batch.append(element)
+        self.batch_counter += 1
+        if self.batch_counter >= self.batch_size:
+            self._flush()
+
+    def _flush(self):
+        if self.batch_counter == 0:
+            return
+
+        with _InsertEmbeddingInOpenSearchSink(self.host, self.port, self.username, self.password, self.embedded_columns) as sink:
```

Comment:
Is there a reason you're instantiating a new instance of the internal class here every time? You wind up creating a new class + client every time

Reply:
I reason I did that because I don't want to bulk push on a stale connection. Given it's a batch push I thought it might be better instead of try catching an exception and reinstantiating a connection after error.
What do you think ?